ENVParser/EnvFile.cs

```diff
@@ -0,0 +1,580 @@
+            FieldModelSpecularAlpha = reader.ReadSingle();
+            FieldModelEmissiveRed = reader.ReadSingle();
+            FieldModelEmissiveGreen = reader.ReadSingle();
+            FieldModelEmissiveBlue = reader.ReadSingle();
+            FieldModelEmissiveAlpha = reader.ReadSingle();
+            Field52 = reader.ReadSingle();
+            Field56 = reader.ReadSingle();
+            Field5A = reader.ReadSingle();
+            Field5E = reader.ReadSingle();
+            FieldModelLightX = reader.ReadSingle();
+            FieldModelLightY = reader.ReadSingle();
+            FieldModelLightZ = reader.ReadSingle();
+            UnusedTextureSection = reader.ReadBytes(188);
+            Field12A = reader.ReadByte();
+            EnableCharacterModelSection = reader.ReadBoolean();
+            CharacterModelAmbientRed = reader.ReadSingle();
+            CharacterModelAmbientGreen = reader.ReadSingle();
+            CharacterModelAmbientBlue = reader.ReadSingle();
+            CharacterModelAmbientAlpha = reader.ReadSingle();
+            CharacterModelDiffuseRed = reader.ReadSingle();
+            CharacterModelDiffuseGreen = reader.ReadSingle();
+            CharacterModelDiffuseBlue = reader.ReadSingle();
+            CharacterModelDiffuseAlpha = reader.ReadSingle();
+            CharacterModelSpecularRed = reader.ReadSingle();
+            CharacterModelSpecularGreen = reader.ReadSingle();
+            CharacterModelSpecularBlue = reader.ReadSingle();
+            CharacterModelSpecularAlpha = reader.ReadSingle();
+            CharacterModelEmissiveRed = reader.ReadSingle();
+            CharacterModelEmissiveGreen = reader.ReadSingle();
+            CharacterModelEmissiveBlue = reader.ReadSingle();
+            CharacterModelEmissiveAlpha = reader.ReadSingle();
+            Field16C = reader.ReadSingle();
+            Field170 = reader.ReadSingle();
+            Field174 = reader.ReadSingle();
+            Field178 = reader.ReadSingle();
+            CharacterModelLightX = reader.ReadSingle();
+            CharacterModelLightY = reader.ReadSingle();
+            CharacterModelLightZ = reader.ReadSingle();
+            Field188 = reader.ReadSingle();
+            ModelNearClip = reader.ReadSingle();
+            ModelFarClip = reader.ReadSingle();
+            EnableFog = reader.ReadBoolean();
+            EnableAmbientFog = reader.ReadBoolean();
+            DisableFog = reader.ReadBoolean();
+            ToggleFogCameraPlane = reader.ReadBoolean();
+            FogStartDistance = reader.ReadSingle();
+            FogEndDistance = reader.ReadSingle();
+            FogRed = reader.ReadSingle();
+            FogGreen = reader.ReadSingle();
+            FogBlue = reader.ReadSingle();
+            FogAlpha = reader.ReadSingle();
+            EnableFloorFog = reader.ReadBoolean();
+            FloorFogStartingHeight = reader.ReadSingle();
+            FloorFogEndHeight = reader.ReadSingle();
+            FloorFogRed = reader.ReadSingle();
```

Comment:
An RGBA class could also save you some space, since those kinds of blocks happen a lot.

Reply:
Yeah, I wanted to separate out the fields that are valid for RGBA vs defining it in the ENVFile itself, so that it in the future more RGBA fields are noted, I can update it one place, rather than looking for the class inthe Env class itself - however happy to hear your suggestions (refer Fields/RGBFieldNameProvider.cs)

Reply:
Hmm, I'm trying to understand the benefit of the validity checker vs. just having something like:

```C#
public class RGBA
{
    public float R;
    public float G;
    public float B;
    public float A;

    public RGBA(BigEndianBinaryReader reader)
    {
        R = reader.ReadSingle();
        G = reader.ReadSingle();
        B = reader.ReadSingle();
        A = reader.ReadSingle();
    }
}
```

...so, in the current class, you could then do...

```C#
    Fog = new RGBA(reader);
    EnableFloorFog = reader.ReadBoolean();
    FloorFogStartingHeight = reader.ReadSingle();
    FloorFogEndHeight = reader.ReadSingle();
    Floor = new RGBA(reader);
```